timemachine/fe/mcgregor.py

```diff
@@ -358,28 +360,35 @@ def mcs(
             filter_fxn,
         )
 
-        total_nodes_visited += mcs_result.nodes_visited
+        visits_left -= mcs_result.nodes_visited
 
-        # If timed out, either due to max_visits or max_cores, raise exception.
-        if mcs_result.timed_out:
-            warnings.warn(
-                f"Reached max number of visits/cores: {len(mcs_result.all_maps)} cores with {mcs_result.nodes_visited} nodes visited. "
-                "Cores may be suboptimal.",
-                MaxVisitsWarning,
-            )
+        total_nodes_visited += mcs_result.nodes_visited
 
         if len(mcs_result.all_maps) > 0:
+            # If we timed out but got cores, throw a warning
+            if mcs_result.timed_out and len(mcs_result.all_maps) < max_cores:
+                warnings.warn(
+                    f"Inexhaustive search: reached max number of visits ({max_visits}) and found only "
+                    f"{len(mcs_result.all_maps)} out of {max_cores} desired cores.",
+                    MaxVisitsWarning,
+                )
             # don't remove this comment and the one below, useful for debugging!
             # print(
             # f"==SUCCESS==[NODES VISITED {mcs_result.nodes_visited} | CORE_SIZE {len([x != UNMAPPED for x in mcs_result.all_maps[0]])} | NUM_CORES {len(mcs_result.all_maps)} | NUM_EDGES {mcs_result.num_edges} | time taken: {time.time()-start_time} | time out? {mcs_result.timed_out}]====="
             # )
             break
+        elif mcs_result.timed_out:
+            # If timed out, either due to max_visits or max_cores, raise exception.
+            raise NoMappingError(
+                f"Exceeded max number of visits/cores - no valid cores could be found: {mcs_result.nodes_visited} nodes visited."
+            )
         # else:
         # print(
         # f"==FAILED==[NODES VISITED {mcs_result.nodes_visited} | time taken: {time.time()-start_time} | time out? {mcs_result.timed_out}]====="
         # )
 
-    assert mcs_result is not None
+    if mcs_result is None:
+        raise NoMappingError("No thresholds evaluated when generating cores")
```

Comment:
[style] I don't know what "no thresholds evaluated" means -- "no layers completed", "timed out before completing first stage", or similar ?

Reply:
The term 'layers' in the code here means the number of layers of recursion. 

Actually means that there is no mapping, will change this to make it clearer and add a test of this logic

Reply:
Fixed up in https://github.com/proteneer/timemachine/pull/1357/commits/fbe77e6b3fd5a81e9d2f9d16bba75d7acf7ee02c to be more clear that this happens when the predicate matrix is empty.